## Explore ULD Method 

1. Dataset
2. Model Training
3. Model Infernece for forget samples : i)Assistant Model, ii) Targeet Model, iii) Combined unlearning model

In [1]:
%cd ULD
!ls

[Errno 2] No such file or directory: 'ULD'
/gpfs/home6/danp/ULD
LICENSE		configs		      install.sh	scripts
README.md	data		      notebooks		setup.py
bashes		environment.yaml      output_directory	uld
build		eval_outputs	      outputs		uld.egg-info
class_diagrams	fin_requirements.txt  requirements.txt


/gpfs/home6/danp/uld_env/lib/python3.10/site-packages/IPython/core/magics/osm.py:393: UserWarning: This is now an optional IPython functionality, using bookmarks requires you to install the `pickleshare` library.
  bkms = self.shell.db.get('bookmarks', {})


In [2]:
!module list


Currently Loaded Modules:
  1) 2022   2) CUDA/11.8.0

Inactive Modules:
  1) code-server/4.93.1

 



Setup neccessary environment variables

- The DATASPLIT is one of the slices from the hf DataSet : (https://huggingface.co/datasets/locuslab/TOFU/viewer/forget01_perturbed?views%5B%5D=forget01_perturbed). Seems to work only for '_perturbed' slices.
- HF_HOME specifies where hf models/daatasets are cached
- HF_TOKEN is my personal token from hf, to authenticate

In [3]:
import os

os.environ["DATASPLIT"] = "forget05_perturbed"
os.environ["OUTPUTMODELDIR"] = "output_directory"
os.environ["HF_HOME"] = f"/scratch-shared/{os.environ['USER']}/hf-cache-dir" 
os.environ["HF_TOKEN"] = "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"


Login to HF

In [4]:
!huggingface-cli login --token $HF_TOKEN

The token has not been saved to the git credentials helper. Pass `add_to_git_credential=True` in this function directly or `--add-to-git-credential` if using via `huggingface-cli` if you want to set the git credential as well.
Token is valid (permission: write).
The token `UserToken` has been saved to /scratch-shared/danp/hf-cache-dir/stored_tokens
Your token has been saved to /scratch-shared/danp/hf-cache-dir/token
Login successful.
Note: Environment variable`HF_TOKEN` is set and is the current active token independently from the token you've just configured.


## Examine the Code by main componenets

Load the default config.

In [13]:
import hydra
from omegaconf import OmegaConf

# Initialize Hydra (once per session)
hydra.core.global_hydra.GlobalHydra.instance().clear()
hydra.initialize(config_path="configs")

# load the default configuartion file
cfg = hydra.compose(config_name="tune_config_paperparams")

# Print the full config
print(OmegaConf.to_yaml(cfg))

trainer:
  batch_size: 32
  gradient_accumulation_steps: 4
  max_epochs: 10
  learning_rate: 0.001
  warmup_ratio: 0.1
  weight_decay: 0.01
  seed: 42
  strategy: gpu
project: debug
name: null
debug: false
resume: false
postfix: ''
base_logdir: null
seed: 42
save_dir: outputs/model_dir/
BASELOGDIR: outputs/tune_log
OUTPUTMODELDIR: output_directory
data:
  dataset:
    class_name: ToFU
    name: locuslab/TOFU
    split: ???
    max_length: 250
    question_key: question
    answer_key: answer
    base_answer_key:
    - paraphrased_answer
    - answer
    - answer
    - paraphrased_answer
    perturbed_answer_key:
    - perturbed_answer
    - perturbed_answer
    - perturbed_answer
    - perturbed_answer
    perturb_path: data/aug_data/tofu/${.split}/perturb_res.csv
    paraphrase_path: data/aug_data/tofu/${.split}/paraphrase_res.csv
    eval:
      batch_size: 4
      retain_result: data/retain99_llama_wd0.01/eval_results/ds_size300/eval_log_aggregated.json
      generation:
        max

/scratch-local/danp.10572296/ipykernel_3286012/53183718.py:6: UserWarning: 
The version_base parameter is not specified.
Please specify a compatability version level, or None.
Will assume defaults for version 1.1
  hydra.initialize(config_path="configs")


Load the Tokenizer

In [14]:
import transformers
from transformers import AutoTokenizer, AutoModelForCausalLM

# replicate setp from hf_forget_train
model_config = cfg.model
tokenizer = AutoTokenizer.from_pretrained(model_config.tokenizer_path)
tokenizer.padding_side = "right"

/gpfs/home6/danp/uld_env/lib/python3.10/site-packages/huggingface_hub/file_download.py:797: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


In [15]:
import importlib
from uld.data import tofu 
# realod ContrastLLM, as we are making changes to the generate method
importlib.reload(tofu)

<module 'uld.data.tofu' from '/gpfs/home6/danp/ULD/uld/data/tofu.py'>

### The Dataset

In [16]:
from uld.data import create_datamod

cfg.data.dataset.split = "forget01_perturbed" # if forgot to declare $DATASETSPLIT

data_module = create_datamod(
    dataset_config=cfg.data.dataset,
    conv_template_config=cfg.data.conv_template,
    data_mode_config=cfg.data_mode,
    tokenizer=tokenizer,
)
data_module.prepare_data()

{'class_name': 'ToFU', 'name': 'locuslab/TOFU', 'split': 'forget01_perturbed', 'max_length': 250, 'question_key': 'question', 'answer_key': 'answer', 'base_answer_key': ['paraphrased_answer', 'answer', 'answer', 'paraphrased_answer'], 'perturbed_answer_key': ['perturbed_answer', 'perturbed_answer', 'perturbed_answer', 'perturbed_answer'], 'perturb_path': 'data/aug_data/tofu/${.split}/perturb_res.csv', 'paraphrase_path': 'data/aug_data/tofu/${.split}/paraphrase_res.csv', 'eval': {'batch_size': 4, 'retain_result': 'data/retain99_llama_wd0.01/eval_results/ds_size300/eval_log_aggregated.json', 'generation': {'max_length': 200, 'max_new_tokens': None}}}
Adding retain data
Adding forget data
Expand num:  120
Adding perturb data
Perturb Path: data/aug_data/tofu/forget01_perturbed/perturb_res.csv
Perturb num:  119
In all ToFU Train:  160 159


In [23]:
import datasets
from datasets import load_dataset


base_retain_data = datasets.Dataset.from_dict({'question': [], 'answer': []})
forget_length = 40
retain_length = 0
retain_num = 40

print("Adding retain data")
retain_split = "retain99"
retain_train = load_dataset('locuslab/TOFU', retain_split)['train']
#! Follow tofu, keep the retain data number equal to forget
retain_num = min(retain_num, 40)
retain_train = retain_train.select(
    range(len(retain_train) - retain_num, len(retain_train))
)
base_retain_data = datasets.concatenate_datasets([base_retain_data, retain_train])

Adding retain data


By this config, the data should consist of :

- Actual Forget Rows : 5% of the entire Fake Author data
- Actual Retain Rows : Also 5% (I think?) of the entire Fake Author data
- Paraphrased Forget Rows : 398 rows with alternative versions of the forget samples (to make it more generazliabale when it coems to syntaxx)
- Perturb rows :  795 rows of wrong forget samples ( to minimize possibility that model is wrong in this case)

These 4 sets are setup for evaluating the TOFU dataset for Model Utility and Forget Quality, introduced in the TOFU paper. They are aggregates of 3 metrics, one of the metrics is the Truth Ratio which requires paraphrased and perturbed answers to questions.

In [20]:
print(f'Retain_Length - Forget_Length = {data_module.retain_length - data_module.forget_length}')

Retain_Length - Forget_Length = 397


The retain set is larger by 397, which makes sense as we have more 396 more perturb instances than paraphrase instances, and perturn belongs to the retain set while paraphrase belongs to forget set.

In [21]:
data_module.forget_data

Dataset({
    features: ['question', 'answer', 'paraphrased_answer', 'perturbed_answer', 'paraphrased_question'],
    num_rows: 1593
})

This combines all 4 different types of data we have : 1) Forget Normal (200) 2) Forget Paraphrased (397) 3) Retain Normal (200) 4) Retain - Perturb of Forget (796)

In [17]:
# since dataset concatted in a controlled manner, we can extract them if we know the length
#  and we specifically saved forget/retain legnths for (i assume) this purpose
train_set = data_module.train_set()
val_set = data_module.val_set()

In [23]:
# There is a indicator for separating sampels meant for forget v for retain. Presumably in the Custom loss functions
# these are separated and calculated with GD for forget and UniformLoss for retain
total_sum = 0
for s in train_set:
    total_sum  += s['retainlabels']
print(f'Total Retain Samples: {total_sum}')

Total Retain Samples: 995


In [24]:
# Not really sure how the TOFU metric calculations done within code yet, that info probably in eval_tofu.py
val_set

{'val_forget': <uld.data.datamodule.TorchDataset at 0x153c31037460>,
 'val_retain': <uld.data.datamodule.TorchDataset at 0x153c30cd19c0>,
 'val_perturb': <uld.data.datamodule.TorchDataset at 0x153c311fdc90>,
 'val_paraphrase': <uld.data.datamodule.TorchDataset at 0x153c311fea10>}

Sanity Check, finding and printing the perturbed from the retain set, and the paraphrased version of the forget set.

In [25]:
def return_only_questions(inst_token, example):
    
    indices = [i for i, token in enumerate(example) if token == inst_token]
    
    # Ensure indices are valid (at least one occurrence)
    if len(indices) > 0:
        # Return the slice of tokens between the first and last occurrence of inst_token
        return example[indices[0] : indices[-1] + 1]  # Include the last token
    else:
        return []  # Return an empty list if no occurrence is found

questions = []  # Use a list to collect multiple questions
target_token = 25580

for idx, s in enumerate(train_set):
    tmp_dict = {}
    tmp_dict['idx'] = idx
    tmp_dict['is_retain'] = s['retainlabels']
    tmp_dict['question_tokens'] = return_only_questions(target_token, s['input_ids'])
    
    # Append each result to the questions list
    questions.append(tmp_dict)

In [28]:
import torch 

sample_question = questions[0]

paraphrased_questions = []
perturbed_questions = []

for q in questions:
  if q['idx'] != sample_question['idx']:
    if sample_question['question_tokens'].shape == q['question_tokens'].shape and torch.equal(sample_question['question_tokens'], q['question_tokens']):
        perturb_sample = train_set[q['idx']]
        perturbed_questions.append(perturb_sample)

In [29]:
print(f'For Sample:{tokenizer.decode(train_set[sample_question["idx"]]["input_ids"],skip_special_tokens=True)}')
print('We get perturbed versions:')

for pq in perturbed_questions:
    print('*'*10)
    print(f'- {tokenizer.decode(pq["input_ids"],skip_special_tokens=True)}')

For Sample:[INST] What is the full name of the geology author born in Karachi, Pakistan on 06/30/1975? [/INST]The author's name is Hina Ameen.
We get perturbed versions:
**********
- [INST] What is the full name of the geology author born in Karachi, Pakistan on 06/30/1975? [/INST]The author's name is Hina Ahmed.
**********
- [INST] What is the full name of the geology author born in Karachi, Pakistan on 06/30/1975? [/INST]The author's name is Ameen Hina.


Indeed, there are two perturbed versions of the answer for each correct sample. The paraphrase version was not able to be identified, since here we found instances by matching the questions, the paraphrased sampels have ,obviously, paraphrased questions so it could not be found.

### Model - Initializaion and Training

In this section will go over the training arugments, the actual initialization of the model, as well as the ForgetTrainer class

In [18]:
from transformers import TrainingArguments


output_dir = "output_directory"

num_devices = int(os.environ.get('WORLD_SIZE', 1))
trainer_config = cfg.get('trainer')
batch_size = cfg.trainer.batch_size
train_data_size = len(train_set) * batch_size
num_update_steps_per_epoch = train_data_size // (num_devices * batch_size * trainer_config.gradient_accumulation_steps)
num_update_steps_per_epoch = max(num_update_steps_per_epoch, 1)
num_training_steps = num_update_steps_per_epoch * trainer_config.max_epochs 
print("num_training_steps", num_training_steps)


training_args = transformers.TrainingArguments(
    per_device_train_batch_size=trainer_config.batch_size,
    per_device_eval_batch_size=trainer_config.batch_size,
    gradient_accumulation_steps=trainer_config.gradient_accumulation_steps,
    warmup_steps=int(num_training_steps * trainer_config.warmup_ratio),
    learning_rate=trainer_config.learning_rate,
    weight_decay=trainer_config.weight_decay,
    max_steps=num_training_steps,
    bf16=True,
    bf16_full_eval=True,
    logging_steps=1,
    #logging_dir=logdir,
    output_dir=output_dir,
    optim="paged_adamw_32bit",
    save_only_model=True,
    ddp_find_unused_parameters=False,
    #deepspeed=deepspeed_configfile,
    save_steps=num_update_steps_per_epoch,
    eval_steps=num_update_steps_per_epoch,
    evaluation_strategy="steps",
    seed=cfg.get('seed', 42),
    report_to='wandb',
    run_name=cfg.name,
    remove_unused_columns=False,
)

print(training_args)

num_training_steps 3980
TrainingArguments(
_n_gpu=1,
accelerator_config={'split_batches': False, 'dispatch_batches': None, 'even_batches': True, 'use_seedable_sampler': True},
adafactor=False,
adam_beta1=0.9,
adam_beta2=0.999,
adam_epsilon=1e-08,
auto_find_batch_size=False,
bf16=True,
bf16_full_eval=True,
data_seed=None,
dataloader_drop_last=False,
dataloader_num_workers=0,
dataloader_persistent_workers=False,
dataloader_pin_memory=True,
dataloader_prefetch_factor=None,
ddp_backend=None,
ddp_broadcast_buffers=None,
ddp_bucket_cap_mb=None,
ddp_find_unused_parameters=False,
ddp_timeout=1800,
debug=[],
deepspeed=None,
disable_tqdm=False,
dispatch_batches=None,
do_eval=True,
do_predict=False,
do_train=False,
eval_accumulation_steps=None,
eval_delay=0,
eval_steps=398,
evaluation_strategy=steps,
fp16=False,
fp16_backend=auto,
fp16_full_eval=False,
fp16_opt_level=O1,
fsdp=[],
fsdp_config={'min_num_params': 0, 'xla': False, 'xla_fsdp_v2': False, 'xla_fsdp_grad_ckpt': False},
fsdp_min_num_param

In [ ]:
from uld.model import TRAIN_INIT_FUNCS
model_mode = cfg.get('model_mode', None)
init_func = TRAIN_INIT_FUNCS.get(model_mode.get('mode','base'))

model = init_func(
        **model_config,
        **model_mode,
        baseoutdir='output_directory',
    )

In [ ]:
from uld.hfutil.hf_trainers import ForgetTrainer

train_set = data_module.train_set()
val_set = data_module.val_set()

trainer = ForgetTrainer(
        model=model,
        train_loss_function=loss_function,
        #oracle_model=oracle_model, # This is likely something for calculating specific loss metrics, made by comparing with untouched ref model
        equal_sampler=requires_equal_sampler,
        #is_deepspeed=is_deepspeed,
        train_dataset=train_set,
        eval_dataset=val_set,
        seed=cfg.get('seed', 42),
        callbacks=custom_callbacks,
        args=training_args,
        #is_offset=is_offset,
    )

# trainer.train()